### PR TITLE
Add `delete+insert` materialization

### PR DESCRIPTION
**Pull request by mdesmet**
## Overview

- Description: Support multiple dbt native incremental strategies (delete+insert, merge (currently deactivated))
- Related issue(s):
- Related code pull request(s):
- Related link(s):

## Checklist

- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] `README.md` updated and added information about my change
- [x] `CHANGELOG.md` updated and added information about my change

### CODE REVIEW

File: dbt/include/trino/macros/materializations/incremental.sql
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:16]: is there any specific reason why the new methods are prefixed with "dbt_trino_" ?

**Comment by findinpath** [comment]
Add a reference towards https://docs.getdbt.com/docs/building-a-dbt-project/building-models/configuring-incremental-models and state the options for the incremental strategy

File: dbt/include/trino/macros/materializations/incremental.sql
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:23]: Is this a valid strategy for Trino at the moment?

https://github.com/trinodb/trino/issues/7708

File: dbt/include/trino/macros/materializations/incremental.sql
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql:16]: I based myself on the snowflake implementation. I think their reasoning is that macro's are global and prefixing it will avoid name clashes from happening, similar like java namespaces.

See [snowflake adapter](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/include/snowflake/macros/materializations/incremental.sql)

File: dbt/include/trino/macros/materializations/incremental.sql
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql:23]: Note that this strategy is currently disabled:

```
  {% set invalid_strategy_msg -%}
    Invalid incremental strategy provided: {{ strategy }}
    Expected one of: 'append', 'delete+insert'
  {%- endset %}
  {% if strategy not in ['append', 'delete+insert'] %}
    {% do exceptions.raise_compiler_error(invalid_strategy_msg) %}
  {% endif %}
```

**Comment by retrry** [comment]
What is needed to push forward this MR?

**Comment by mdesmet** [comment]
I actually had a stab at the integration test for this feature with postgres but the query is not supported

```
delete from "postgresql"."test16528192466294309628_test_incremental_delete_insert"."incremental"
            where (
                id) in (
                select (id)
                from "postgresql"."test16528192466294309628_test_incremental_delete_insert"."incremental__dbt_tmp"
            )
```

```
io.trino.spi.TrinoException: Unsupported delete
	at io.trino.plugin.jdbc.DefaultJdbcMetadata.beginDelete(DefaultJdbcMetadata.java:733)
	at io.trino.spi.connector.ConnectorMetadata.beginDelete(ConnectorMetadata.java:664)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.beginDelete(ClassLoaderSafeConnectorMetadata.java:684)
	at io.trino.metadata.MetadataManager.beginDelete(MetadataManager.java:1002)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.createWriterTarget(BeginTableWrite.java:265)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:201)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:104)
	at io.trino.sql.planner.plan.TableFinishNode.accept(TableFinishNode.java:106)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:37)
	at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:21)
	at io.trino.sql.planner.plan.PlanVisitor.visitOutput(PlanVisitor.java:49)
	at io.trino.sql.planner.plan.OutputNode.accept(OutputNode.java:83)
	at io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith(SimplePlanRewriter.java:31)
	at io.trino.sql.planner.optimizations.BeginTableWrite.optimize(BeginTableWrite.java:89)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:239)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:216)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:211)
	at io.trino.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:477)
	at io.trino.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:458)
	at io.trino.execution.SqlQueryExecution.start(SqlQueryExecution.java:399)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:243)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:143)
	at io.trino.$gen.Trino_379____20220517_201455_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
	Suppressed: java.lang.Exception: Current plan:
                Output[rows]
                │   Layout: [rows:bigint]
                └─ TableCommit[[]]
                   │   Layout: [rows:bigint]
                   └─ LocalExchange[SINGLE] ()
                      │   Layout: [partialrows:bigint, fragment:varbinary]
                      └─ RemoteExchange[GATHER]
                         │   Layout: [partialrows:bigint, fragment:varbinary]
                         └─ Delete[[]]
                            │   Layout: [partialrows:bigint, fragment:varbinary]
                            └─ FilterProject[filterPredicate = "expr"]
                               │   Layout: [field:bigint]
                               └─ Project[]
                                  │   Layout: [field:bigint, expr:boolean]
                                  └─ SemiJoin[id = id_0][$hashvalue, $hashvalue_6]
                                     │   Layout: [id:integer, field:bigint, $hashvalue:bigint, expr:boolean]
                                     │   Distribution: REPLICATED
                                     │   dynamicFilterId: df_554
                                     ├─ ScanFilterProject[table = postgresql:test16528192466294309628_test_incremental_delete_insert.incremental test16528192466294309628_test_incremental_delete_insert.incremental columns=[id:integer:int4, $update_row_id:bigint:bigint], filterPredicate = true, dynamicFilters = {"id" = #df_554}]
                                     │      Layout: [id:integer, field:bigint, $hashvalue:bigint]
                                     │      $hashvalue := combine_hash(bigint '0', COALESCE("$operator$hash_code"("id"), 0))
                                     │      field := $update_row_id:bigint:bigint
                                     │      id := id:integer:int4
                                     └─ LocalExchange[SINGLE] ()
                                        │   Layout: [id_0:integer, $hashvalue_6:bigint]
                                        └─ RemoteExchange[REPLICATE]
                                           │   Layout: [id_0:integer, $hashvalue_7:bigint]
                                           └─ ScanProject[table = postgresql:test16528192466294309628_test_incremental_delete_insert.incremental__dbt_tmp test16528192466294309628_test_incremental_delete_insert.incremental__dbt_tmp columns=[id:integer:int4]]
                                                  Layout: [id_0:integer, $hashvalue_8:bigint]
                                                  $hashvalue_8 := combine_hash(bigint '0', COALESCE("$operator$hash_code"("id_0"), 0))
                                                  id_0 := id:integer:int4

		at io.trino.sql.planner.optimizations.BeginTableWrite.optimize(BeginTableWrite.java:95)
		... 12 more
```

Next i will try with delta lake + minio. I will probably create a separate PR to add delta + minio first.

**Comment by findinpath** [comment]
@mdesmet 

Today DELETE on rdbms connectors is only supported when the predicate can be entirely pushed down to the connector.
Also neither DELETE or UPDATE support using subqueries as of now.

**Comment by mdesmet** [comment]
> @mdesmet
> 
> Today DELETE on rdbms connectors is only supported when the predicate can be entirely pushed down to the connector. Also neither DELETE or UPDATE support using subqueries as of now.

Indeed, that was also my conclusion. 

I did a basic test with the Delta connector and there DELETE with subqueries are supported.

```
CREATE TABLE test_delta.test.test AS 
SELECT * from tpch.sf1.customer;

DELETE FROM test_delta.test.test WHERE custkey
 IN (SELECT custkey from test_delta.test.test where custkey % 15 <> 0)
```

**Comment by hovaesco** [comment]
Probably we would need to switch from PostgreSQL to Minio for testing.

**Comment by mdesmet** [comment]
> Probably we would need to switch from PostgreSQL to Minio for testing.

Yes indeed, will post a PR with delta + minio + hms support later today.

File: dbt/include/trino/macros/materializations/incremental/merge.sql
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental/merge.sql:4]: `unique_key` is not required. In that case it actually behaves like `append`. This behavior is tested in following tests:

- test__empty_str_unique_key
- test__empty_unique_key_list

It just behaves exactly as in OOTB dbt (snowflake).

File: tests/functional/adapter/materialization/test_incremental_delete_insert.py
Comment (mdesmet) [reply tests/functional/adapter/materialization/test_incremental_delete_insert.py:136]: 🥱

File: tests/functional/adapter/materialization/test_incremental_delete_insert.py
Comment (mdesmet) [reply tests/functional/adapter/materialization/test_incremental_delete_insert.py:164]: I think i overcomplicated it a bit. I will remove the abstract method and just provide a `project_config_update` method per catalog type.

File: dbt/include/trino/macros/materializations/incremental/incremental.sql
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental/incremental.sql:25]: It's defintely correct what you're saying but the last branch would be an else then. I think this way is more explicit and does not duplicate the default configuration from `config.get("incremental_strategy", default="append")`.

File: dbt/include/trino/macros/materializations/incremental/incremental.sql
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental/incremental.sql:1]: I think it is a little cleaner to split it to keep the logic in incremental light.

File: tests/functional/adapter/materialization/test_incremental_delete_insert.py
Comment (mdesmet) [reply tests/functional/adapter/materialization/test_incremental_delete_insert.py:45]: this is coming straight form the standard dbt tests. I think as this is a supporting model (model for expected values check) dbt prefers to exclude it from the project_config_update to only keep relevant project configuration there that users should look at if they implement an incremental model.

https://github.com/dbt-labs/dbt-core/blob/94a7cfa58df3df5d3b07b619c0510fead9195700/tests/adapter/dbt/tests/adapter/incremental/test_incremental_unique_id.py#L218-L222

File: README.md
Comment (hovaesco) [inline README.md:196]: Can this be somehow rephrased? This strategy deletes records and insert new ones under the hood. So from the end user perspective it's upsert but from engine perspective it's not.

File: tests/functional/adapter/materialization/test_incremental_delete_insert.py
Comment (hovaesco) [reply tests/functional/adapter/materialization/test_incremental_delete_insert.py:45]: thanks

**Review by hovaesco — APPROVED** [review]
LGTM % some README comments and please update CHANGELOG and change commit message to reflect that `delete+insert` strategy is added. 

Let's wait for @findinpath review before merging.

File: dbt/include/trino/macros/materializations/incremental/incremental.sql
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental/incremental.sql:21]: i don't quite follow why do we have this branch now (as written, `MERGE` is at the moment) not yet supported. Why not creating a follow-up PR when a Trino version with `MERGE` support lands?

File: dbt/include/trino/macros/materializations/incremental/incremental.sql
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental/incremental.sql:27]: Is leaving `dbt_trino_` prefix done on purpose?

File: dbt/include/trino/macros/materializations/incremental/merge.sql
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental/merge.sql:27]: i'm wondering whether you missed 

```
.include(database=true, schema=true)
```

like it is done for `append`

https://github.com/starburstdata/dbt-trino/pull/62/files#diff-a5aa2b035a51111d752170921f513337ce6af7d4231b586252df7b0a7dbf71b6R30

File: tests/functional/adapter/materialization/test_incremental_delete_insert.py
Comment (findinpath) [inline tests/functional/adapter/materialization/test_incremental_delete_insert.py:49]: nit: can be done with `VALUES ... AS t(....)` as well

File: dbt/include/trino/macros/materializations/incremental/incremental.sql
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental/incremental.sql:27]: I think they did that in the dbt-snowflake adapter to _prematurely_ avoid name clashes with internal dbt core macros. Remember all macros are loaded onto the same Jinja environment and name clashes are resolved based on load order.

File: tests/functional/adapter/materialization/test_incremental_delete_insert.py
Comment (findinpath) [inline tests/functional/adapter/materialization/test_incremental_delete_insert.py:133]: why not `date` ?

File: dbt/include/trino/macros/materializations/incremental/incremental.sql
Comment (findinpath) [reply dbt/include/trino/macros/materializations/incremental/incremental.sql:27]: I'm advocating that we should have a consistent way of defining macros in the adapter. 
Any form that keeps the existing functionality is fine, but should be consistent across the codebase of the adapter. (obviously this should be a follow-up PR).

Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental/incremental.sql:27]: Let's go for the removal of the prefix for now

File: tests/functional/adapter/materialization/test_incremental_delete_insert.py
Comment (mdesmet) [reply tests/functional/adapter/materialization/test_incremental_delete_insert.py:49]: This is an override from the original test in dbt's code. I prefer to keep it as close to the original as possible in case we need to compare it later on.

File: tests/functional/adapter/materialization/test_incremental_delete_insert.py
Comment (mdesmet) [reply tests/functional/adapter/materialization/test_incremental_delete_insert.py:133]: Good point!

File: dbt/include/trino/macros/materializations/incremental/merge.sql
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental/merge.sql:27]: This is not required, it is actually already generated using database and schema. In the line after this one we can see:

`drop table if exists {{ tmp_relation }};`

If it would be necessary we should include it there too, otherwise the table would not be found.

Note that snowflake also doesn't do that.

**Review by hovaesco — APPROVED** [review]

**Comment by findinpath** [comment]
Please rebase your changes on `master`

File: dbt/include/trino/macros/materializations/incremental.sql
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:24]: We've discussed this already.

I see `get_append_sql` and `trino__get_delete_insert_merge_sql` which are namewise not consistent.

I'd be in favor of prefixing with `trino__` until we get consistent macro names across all the codebase.

File: dbt/include/trino/macros/materializations/incremental.sql
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:40]: Do we need to eventually take into consideration quoting the column names here?
If yes, could you please add a corresponding test case ?

Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:40]: most of the statements here use lowercase for the reserved sql keywords.
pls change `IN` `SELECT and `FROM` accordingly.

File: dbt/include/trino/macros/materializations/incremental.sql
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:20]: is `_merge` needed in the macro name?

File: dbt/include/trino/macros/materializations/incremental.sql
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql:24]: I suggest that the `trino__` prefix should only be used for adapter macros that are dispatched in core dbt. The `get_append_sql` macro is not part of core dbt while `get_delete_insert_merge_sql` is. See https://github.com/dbt-labs/dbt-core/blob/1071a4681df91633301fdf23e34de819b66fbeb7/core/dbt/include/global_project/macros/materializations/models/incremental/merge.sql#L51-L53

File: dbt/include/trino/macros/materializations/incremental.sql
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql:40]: I think the unique_key is supplied by the user here. It's up to the user to supply a quoted or unquoted identifier.

File: dbt/include/trino/macros/materializations/incremental.sql
Comment (findinpath) [reply dbt/include/trino/macros/materializations/incremental.sql:24]: How do we know whether a certain macro is not going to be part of core dbt?

In any case keep this in mind for the eventual follow-up task of having consistent naming for the macros.

**Review by findinpath — APPROVED** [review]

**Pull request merged**
